Minor enhancement to TCacheEntry

iwubcode opened the pull request:
This is pulled from @phire 's original #3880 PR.

This adds two helper functions to deal with the TCacheEntry's format.  In particular, this creates a single place where we can change the bitmasking of the format to get the texture format enumeration value.  When we go to change the mask for Hybrid XFB, this will make less code modifications needed for that change.

Additionally, this will help with some errors seen in #5718 .

iwubcode commented:
I'd be interested in hearing other opinions about the bitset approach (including @phire 's original reasoning for it over individual bools).  @mimimi085181 wasn't a huge fan of this approach and I wondered if others shared his concern.

I'd be ok dropping the bitset and modifying this PR (or opening another) with just my TCacheEntry format changes if that's what we decide on.

lioncash commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -107,6 +120,11 @@ class TextureCacheBase
     u32 GetNumLevels() const { return texture->GetConfig().levels; }
     u32 GetNumLayers() const { return texture->GetConfig().layers; }
     AbstractTextureFormat GetFormat() const { return texture->GetConfig().format; }
+    u32 InMemoryFormat() const { return format & 0xff; }
```
Shouldn't this be `0xF` instead of `0xFF`?

phire replied:
Part of the original reason for this change was to allow easy changing of the mask in just one place. 

My version of hybrid introduced an in memory format which was 6 bits long. I'm not sure what the current status of that is.

iwubcode replied:
Phire basically summed it up but yes that in memory format for the hybrid texture is still there in the hybrid PR.  And it seems like there are other formats that are 6bits as well.  Just look the enumerations for `TextureFormat`

lioncash replied:
Ah OK. Thanks for clarifying

phire replied:
To clarify, there have always been 6 bit formats, but the two extra bits were only used during efb copies.

Until hybrid, we could guarantee that tcache entries would be using one of the 4bit formats.

iwubcode replied:
Switched it back to 0xf, the change to a different mask should probably happen in the hybrid PR

lioncash commented on Source/Core/VideoBackends/OGL/TextureCache.cpp:
```diff
@@ -388,12 +388,13 @@ void TextureCache::ConvertTexture(TCacheEntry* destination, TCacheEntry* source,
   glUniform1i(s_palette_buffer_offset_uniform[format], buffer.second / 2);
-  glUniform1f(s_palette_multiplier_uniform[format], (source->format & 0xf) == 0 ? 15.0f : 255.0f);
+  glUniform1f(s_palette_multiplier_uniform[format],
+              (source->InMemoryFormat()) == 0 ? 15.0f : 255.0f);
```
The parentheses around the function call aren't necessary.

lioncash commented on Source/Core/VideoBackends/D3D/TextureCache.cpp:
```diff
@@ -144,7 +144,7 @@ void TextureCache::ConvertTexture(TCacheEntry* destination, TCacheEntry* source,
 
   // TODO: Add support for C14X2 format.  (Different multiplier, more palette entries.)
-  float params[4] = {(source->format & 0xf) == GX_TF_I4 ? 15.f : 255.f};
+  float params[4] = {(source->InMemoryFormat()) == GX_TF_I4 ? 15.f : 255.f};
```
Parentheses around the function call aren't necessary.

iwubcode commented:
While I didn't expect any changes, I compared this against master with [fifo_comparer](https://github.com/iwubcode/fifo_comparer) and confirmed there were no changes.

phire commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -92,11 +87,29 @@ class TextureCacheBase
 
+    enum FlagPos
+    {
+      EFB_COPY_FLAG_POS = 0,
```
I think you might be better off using a [Bit field](http://en.cppreference.com/w/cpp/language/bit_field) rather than a bitset and this enum.

With a Bit field, it would be stupidly easy and sane to put the flags and the format into a single u32.

iwubcode replied:
Eh..someone else on IRC mentioned that too. I'm not a huge fan of bitfield.  I can't remember where it was but something I had to do on Vertex Rounding really turned me off to them.  They feel archaic and can be painful to update when dealing with many attributes (not to mention dealing with padding and such).  I'll look at them some more though and see if I can get on board with them.

I can't speak to bitset much either to be fair, I've only used it a few times and this was my first time using it for flags.  I think it's more eloquent than bitfields but figure they aren't always interchangeable.

Honestly, I'm leaning more toward going back to bools.  Did you do some profiling in your initial PR?  Why did you decide to move to a single value for format + flags?

iwubcode commented:
After some discussion on IRC, I've decided to keep the original bool values and forgo the optimization using the bitset or bitfield.  If we do profiling in the future, we can reevaluate whether those are advantageous.  For now, I think it doesn't make sense to complicate the code for an unknown advantage.

mimimi085181 commented:
Hmm, i think i found a bug, in master...

If an efb copy is actually a paletted texture, when loading it, Dolphin applies the palette of the texture and returns the "new" texture. The efb copies always get the format without palette, and that is taken over to the "new" texture. When loading this texture again, the format is compared against the full format with the palette, so it has to fail.

Maybe your FullFormat() function should return the format with the palette, independent whether it's actually stored or not in the format variable. If InMemoryFormat() and FullFormat() functions are used consequently, it could also just store the format without palette, and you could save that bitmask.

iwubcode commented:
@mimimi085181 - this has meant to be a fast PR to be merged in.  I am happy to take a look at this in a separate PR.

> If you change this, there could be changes on the fifo comparer, just stating the obvious.

All fifo comparer does is use the output from Dolphin, so I doubt there would be changes, unless I'm misunderstanding you.

> Can you move the format conversion into TexDecoder_GetBlockWidthInTexels() and friends, instead of passing it as argument all the time?

When looking into this, I thought about adding a function.  I plan on doing that in a separate PR.  I'm trying to implement smaller PRs so they get merged more quickly...

mimimi085181 commented:
Well, why not store a new bool, is_paletted_texture, or make a function that checks against the 3 paletted formats? This way we can save all the format conversion. I guess the new format for hybrid xfb is also just adding one or more bools?

phire commented:
Go with the `isPalettedTexture` function. You don't want another bool
encoding information which is already encoded.

iwubcode commented:
Ok a few changes

- Switched to FullFormat in ApplyPaletteToEntry to mimic original code
- Changed "InMemoryFormat" to "TextureFormat" as there was another function dealing with 'in memory' formats, so this makes the name more appropriate
- Slight optimization in the texture cache's to save off the texture format, previously it was being computed multiple times.


I'll open a PR that creates a new 'isPalettedTexture' function as soon as this is merged.

iwubcode commented:
Closing in favor of #5849
